### DISCUSSION TITLE
cache: no-store not working

### DISCUSSION THREAD

**Discussion by alifares10** [vercel/next.js#51771] [Help] [answered]
### Summary

I have a fetch to get data from database but the cache: no-store is not working cause I'm not getting the latest data from the database I'm only getting the data before the deploy, i have the same fetch but in a dynamic route and it works well , i also tried revalidate and tried deleting .next folder and then rebuilding the project again but still no luck .
i also would like to note that the same thing happened with axios also

### Additional information

```js
this is the fetch:

`useEffect(() => {
    const fetchProjects = async () => {
      const res = await fetch("/api/project", {
        cache: "no-store",
      });
      const data = await res.json();
      setAllProjects(data);
    };

    fetchProjects();
  }, []);`
```

**Comment by rasif-sahl** [comment, marked as answer]
1. In your API endpoint (/api/project), make sure you are setting appropriate Cache-Control headers to prevent caching. You can set it to no-store to indicate that the response should not be stored in any cache. Here's an example of how you can set the header using Express.js:
`res.setHeader('Cache-Control', 'no-store');`
2. To bypass any caching issues on the client side, you can add a unique query parameter to the fetch URL. This ensures that each request is treated as a separate URL, preventing any caching. You can use a timestamp or a random value as the query parameter. Here's an example of how you can modify your fetch code to include a unique query parameter:

```useEffect(() => {
  const fetchProjects = async () => {
    const timestamp = new Date().getTime(); // Generate a unique timestamp
    const url = /api/project?timestamp=${timestamp}; // Append timestamp as query parameter
    const res = await fetch(url, {
      cache: 'no-store',
    });
    const data = await res.json();
    setAllProjects(data);
  };
  fetchProjects();
 }, []);
```

3. Ensure that you are not accidentally rendering stale data on the server side. If you are using server-side rendering, make sure that the server is properly fetching the latest data from the database during each request.

**Reply by alifares10** [reply]
this is my api endpoint (im not using express) 
```
export const GET = async (req) => {
  try {
    await connectToDB();
    const projectSchema = await Projects.find({}).populate("_id");
   
    return new Response(JSON.stringify(projectSchema), { status: 201 });
  } catch (error) {
    new Response("Failed to fech Project", { status: 500 });
  }
};
```

how do i set the headers ?

**Reply by alifares10** [reply]
the second option worked thanks

**Reply by rasif-sahl** [reply]
Sure great.
